src/main.rs

```diff
@@ -127,53 +166,97 @@ async fn run() -> Result<(), Box<dyn std::error::Error>> {
-    debug!("Received CLI arguments: {:?}", std::env::args());
+    info!(root_logger, "Starting PokeBot!");
+    debug!(root_logger, "Received CLI arguments"; "args" => ?std::env::args());
 
-    if bot_args.local {
+    if local {
         let name = bot_args.names[0].clone();
-        let id = bot_args.ids.expect("identies should exists")[0].clone();
-
-        let disconnect_cb = Box::new(move |_, _, _| {});
+        let identity = bot_args.ids.expect("identies should exists")[0].clone();
 
         let bot_args = MusicBotArgs {
             name,
-            name_index: 0,
-            id_index: 0,
+            master: None,
             local: true,
             address: bot_args.address.clone(),
-            id,
+            identity,
             channel: String::from("local"),
             verbose: bot_args.verbose,
-            disconnect_cb,
+            logger: root_logger,
         };
-        MusicBot::new(bot_args).await.1.await;
+        MusicBot::new(bot_args).await;
+
+        ctrl_c.await??;
     } else {
         let domain = bot_args.domain.clone();
         let bind_address = bot_args.bind_address.clone();
-        let (bot, fut) = MasterBot::new(bot_args).await;
+        let bot_name = bot_args.master_name.clone();
+        let bot_logger = root_logger.new(o!("master" => bot_name.clone()));
+        let bot = MasterBot::new(bot_args, bot_logger).await;
+
+        let weak_bot = bot.downgrade();
 
-        thread::spawn(|| {
+        let clogger = root_logger.clone();
+        thread::spawn(move || {
             let web_args = web_server::WebServerArgs {
                 domain,
                 bind_address,
-                bot,
+                bot: weak_bot,
             };
-            if let Err(e) = web_server::start(web_args) {
-                error!("Error in web server: {}", e);
+            if let Err(e) = web_server::start(web_args, clogger.new(o!("webserver" => bot_name))) {
+                error!(clogger, "Error in web server"; "error" => %e);
             }
         });
 
-        fut.await;
-        // Keep tokio running while the bot disconnects
-        tokio::time::delay_for(Duration::from_secs(1)).await;
+        #[cfg(unix)]
+        tokio::select! {
+            res = ctrl_c => {
+                res??;
+                info!(root_logger, "Received signal, shutting down"; "signal" => "SIGINT");
+            }
+            _ = sigterm => {
+                info!(root_logger, "Received signal, shutting down"; "signal" => "SIGTERM");
+            }
+            _ = sighup => {
+                info!(root_logger, "Received signal, shutting down"; "signal" => "SIGHUP");
+            }
+            _ = sigquit => {
+                info!(root_logger, "Received signal, shutting down"; "signal" => "SIGQUIT");
+            }
+        };
+
+        #[cfg(windows)]
+        ctrl_c.await??;
+
+        bot.send(Quit(String::from("Stopping")))
+            .await
+            .unwrap()
+            .unwrap();
```

Comment:
We should look into anyhow or something so we can ?? from everywhere :-)

Reply:
Another PR.